torch/_inductor/ir.py

```diff
@@ -3830,13 +3830,23 @@ def codegen(self, wrapper):
             kernel, configs, self.kwargs
         )
 
+        args = self.codegen_kwargs()
+        if V.graph.cpp_wrapper:
+            from torch._higher_order_ops.triton_kernel_wrap import kernel_side_table
+
+            # in C++ wrapper, we don't pass constexpr args, as they don't
+            # get added asparameters to the PTX code compiled from the
+            # user-defined Triton kernel (only non-constexpr args do)
+            kernel = kernel_side_table.get_kernel(self.kernel_idx)
+            args = [arg for i, arg in enumerate(args) if i not in kernel.constexprs]
```

Comment:
I am not sure whether this is guaranteed to be correct. Your 'args' comes from kwargs, which are not guaranteed to be the right order, so I think you need to sort them to the correct order expected by the kernel. I did this somewhere else for the C++ codegen as well.

Reply:
Before this PR, in the CPP wrapper mode, `self.codegen_kwargs()` used to be passed directly to the `CppWrapperCodeGen.generate_user_defined_triton_kernel` method here:

https://github.com/pytorch/pytorch/blob/85aa3723749e0d06aa5fd34215b9b93529a60995/torch/_inductor/codegen/wrapper.py#L1771

which, in turn, passes the `args`, without modifying the order, to the `CppWrapperCodeGen.generate_kernel_call` method here:

https://github.com/pytorch/pytorch/blob/85aa3723749e0d06aa5fd34215b9b93529a60995/torch/_inductor/codegen/wrapper.py#L2570-L2572

finally, the `WrapperCodeGen.wrapper_kernel_call` is called:

https://github.com/pytorch/pytorch/blob/85aa3723749e0d06aa5fd34215b9b93529a60995/torch/_inductor/codegen/wrapper.py#L980

Looking into the chain of calls above, I don't see reordering of the args along the code path. Am I missing something?

Importantly, when compiling a Triton kernel and skipping the `tl.constexpr` parameters, Triton compiler seems to preserve the order of the remaining ones in the emitted PTX code.

Reply:
I think, similar thing is done here for similar purposes?

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/triton_heuristics.py#L310-L314

Reply:
`self.fn.arg_names` are already sorted becayse they come from the definition

Reply:
@oulgen Ok, I think the ordering you're talking about is already happening in `self.codegen_kwargs`:

https://github.com/pytorch/pytorch/blob/85aa3723749e0d06aa5fd34215b9b93529a60995/torch/_inductor/ir.py#L3636-L3642

So we should be good.

Reply:
For the C++ wrapper, the sorting happen on
https://github.com/pytorch/pytorch/blob/fd1a01a393cfe26205795a7fae48ff4e49b5a564/torch/_inductor/ir.py#L3637-L3642

which is right before this, so i think we are all good since you're operating over already sorted list